das/archiving_storage_service.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package das
+
+import (
+	"context"
+	"errors"
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/util/arbmath"
+	"sync"
+	"time"
+)
+
+var ErrArchiveTimeout = errors.New("Archiver timed out")
+
+type ArchivingStorageService struct {
+	inner                  StorageService
+	archiveTo              StorageService
+	archiveChan            chan []byte
+	archiveChanClosed      bool
+	archiveChanClosedMutex sync.Mutex
+	hardStopFunc           func()
+	stoppedSignal          chan interface{}
+	archiverError          error
+}
+
+func NewArchivingStorageService(
+	ctx context.Context,
+	inner StorageService,
+	archiveTo StorageService,
+	archiveExpirationSeconds uint64,
+) (*ArchivingStorageService, error) {
+	archiveChan := make(chan []byte, 256)
+	hardStopCtx, hardStopFunc := context.WithCancel(ctx)
+	ret := &ArchivingStorageService{
+		inner:         inner,
+		archiveTo:     archiveTo,
+		archiveChan:   archiveChan,
+		hardStopFunc:  hardStopFunc,
+		stoppedSignal: make(chan interface{}),
+	}
+
+	go func() {
```

Comment:
`StopWaiter` doesn't quite work here, because in `Close` we want to do a `select` where we wait for the server to stop and also for something else. But `StopWaiter` doesn't provide a channel to signal completion. (That's because `StopWaiter` uses `sync.WaitGroup` which doesn't provide a completion channel.)

(`StopWaiter` has a channel that signals that a stop was requested, but not one that signals that the stop has completed.)

I considered rewriting `StopWaiter` to provide a completion channel, but that seemed more difficult.